### ISSUE TITLE
Phantom error when a `non_local_definitions` warning is emitted

### ISSUE THREAD

**Issue by its-the-shrimp** [rust-lang/rust-analyzer#17450]
rust-analyzer reports an error in the code snippet below at `*k  = Some(v)` saying `cannot mutate immutable variable "k"`.
But only when there's a `non_local_definitions` warning emitted above it.
`cargo check` doesn't report such an error.
The type of `k` inferred by rust-analyzer is `<&mut Trie<T> as DeserializeSeed<'de>>::Value` instead of the expected `&mut Option<T>`, which might cause the phantom error to appear.

**rust-analyzer version**: `rust-analyzer 0.3.2002-standalone (6b8b8ff4c 2024-06-14)`

**rustc version**: `rustc 1.81.0-nightly (59e2c01c2 2024-06-17)`

**editor or extension**: Vim (with the LSP installed by Mason & configured by rust-tools.nvim)

**relevant settings**:

**repository link (if public, optional)**:

**code snippet to reproduce**: 
```rust
use std::{fmt::Formatter, marker::PhantomData};
use serde::{de::{DeserializeSeed, MapAccess, Visitor}, Deserialize, Deserializer};

pub struct Trie<T> {
    _value: Option<T>,
    _submaps: Vec<(u32, Trie<T>)>,
}

impl<'de, T: Deserialize<'de>> Deserialize<'de> for Trie<T> {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        impl<'trie, 'de, T> DeserializeSeed<'de> for &'trie mut Trie<T> {
            type Value = &'trie mut Option<T>;

            fn deserialize<D: Deserializer<'de>>(self, d: D) -> Result<Self::Value, D::Error> {
                todo!()
            }
        }

        struct TrieBuilder<T>(PhantomData<T>);

        impl<'de, T: Deserialize<'de>> Visitor<'de> for TrieBuilder<T> {
            type Value = Trie<T>;

            fn expecting(&self, _: &mut Formatter) -> std::fmt::Result {
                todo!()
            }

            fn visit_map<A: MapAccess<'de>>(self, mut map: A) -> Result<Self::Value, A::Error> {
                let mut trie = Trie { _value: None, _submaps: vec![] };
                while let Some((k, v)) = map.next_entry_seed(&mut trie, PhantomData::<T>)? {
                    *k = Some(v);
                }
                Ok(trie)
            }
        }

        deserializer.deserialize_map(TrieBuilder(PhantomData))
    }
}

fn main() {}
```

**Comment by flodiebold** [comment]
The whole reason for this lint is that RA won't support such definitions, so inference problems when you write such definitions are expected.

**Comment by davidbarsky** [comment]
For additional context, the reason that rust-analyzer won't support such definitions is that it'd require it to look at function bodies when doing global searches (which trait solving is), and that would be unacceptably slow.
